R/fcm.R

```diff
@@ -141,57 +140,50 @@ fcm.dfm <- function(x, context = c("document", "window"),
     ordered <- check_logical(ordered)
     tri <- check_logical(tri)
     
-    margin <- colSums(x)
+    attrs <- attributes(x)
     if (!nfeat(x)) {
-        result <- new("fcm", as(make_null_dfm(), "dgCMatrix"), count = count,
-                      context = context, margin = margin, weights = 1, tri = tri)
+        result <- build_fcm(
+            as(make_null_dfm(), "dgCMatrix"), featnames(x), 
+            count = count, context = context, margin = featfreq(x), 
+            weights = 1, tri = tri,
+            meta = attrs[["meta"]])
+        
         return(result)
     }
 
     if (context != "document")
         stop("fcm.dfm only works on context = \"document\"")
-
+    
+    if (count == "weighted")
+        stop("Cannot have weighted counts with context = \"document\"")
     if (count == "boolean") {
-        temp <- x > 1
+        m <- Matrix::colSums(x > 1)
         x <- dfm_weight(x, "boolean")
-    } else if (count == "frequency") {
-        temp <- x
-        temp@x <- choose(temp@x, 2)
     } else {
-        stop("Cannot have weighted counts with context = \"document\"")
+        m <- Matrix::colSums(x)
     }
+    temp <- Matrix::crossprod(x)
+    
+    # correct self-cooccuerace
+    if (count == "boolean") {
+        Matrix::diag(temp) <- m
+    } else {
+        Matrix::diag(temp) <- (Matrix::diag(temp) - m) / 2    
+    }
+    if (tri) 
+        temp <- Matrix::triu(temp)
 
-    # compute co_occurrence of the diagonal elements
-    sum_col <- colSums(temp) # apply(temp, MARGIN = 2, sum)
-    feature <- sum_col >= 1
-    index_diag <- which(feature)
-    length_feature <- length(feature)
-    temp <- Matrix::sparseMatrix(i = index_diag,
-                                 j = index_diag,
-                                 x = sum_col[feature],
-                                 dims = c(length_feature, length_feature))
-
-    result <- Matrix::crossprod(x)
-    diag(result) <- 0
-    result <- result + temp
-    result <- result[rownames(result), colnames(result)]
-
-    # discard the lower diagonal if tri == TRUE
-    if (tri) result <- triu(result)
-
-    # create a new feature context matrix
-    result <- new("fcm", as(result, "dgCMatrix"), count = count,
-                  context = context, margin = margin, weights = 1,
-                  tri = tri)
-    set_fcm_slots(result) <- attributes(x)
-    set_fcm_dimnames(result) <- list(rownames(result), colnames(result))
+    result <- build_fcm(
```

Comment:
I see how this is consistent now with the other constructors, but I feel they could use more documentation. For instance here you are passing the attributes that were formerly slots, to become object metadata. The use of `...` however is not very clear in the documentation.

The more we document, the more I - and you, in six months! - will have to scratch our heads trying to figure out how these build, rebuild, and upgrade functions work.

Last week for instance I was trying to figure out how quickly to construct a tokens object from an externally passed list of integers and the vocabulary ("types") object. I could call the `build_tokens()` but it also needed a very specific kind of docvars with the `docname_`, `docid_` columns etc. I figured it out only through reading the code and a bit of trial and error.

Last week I

Reply:
Improved documentation.